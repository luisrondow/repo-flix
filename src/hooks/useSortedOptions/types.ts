import { TECHS, SORT_OPTIONS } from '../../utils/constants'

export type SortedOptions = {
  tech: (typeof TECHS)[number]
  sort: (typeof SORT_OPTIONS)[number]['value']
}[]
