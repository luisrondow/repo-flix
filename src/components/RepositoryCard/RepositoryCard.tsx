import { useState } from 'react'

import { RepositoryCardProps } from './RepositoryCard.types'
import StarIcon from '../StarIcon'

const RepositoryCard = (props: RepositoryCardProps) => {
  const {
    repository: { id, name, image },
    onRepositoryClick,
    isBookmarked,
    type,
  } = props

  const isListCard = type === 'list'

  const [isBookmarkedState, setIsBookmarkedState] = useState(isBookmarked)
  const [isCardHovered, setIsCardHovered] = useState(false)

  return (
    <div
      data-testid={`repository-card-${id}`}
      className={`card relative ${
        isListCard ? 'h-56 w-112' : 'h-36 w-72'
      } flex-shrink-0 flex-grow basis-112 cursor-pointer snap-center scroll-mx-96 transition-transform hover:z-10 hover:scale-115 hover:shadow-xl`}
      onMouseEnter={() => isListCard && setIsCardHovered(true)}
      onMouseLeave={() => isListCard && setIsCardHovered(false)}
    >
      <img src={image} alt={name} className="h-full w-full" onClick={onRepositoryClick} />
      {isCardHovered || !isListCard ? (
        <button
          data-testid="bookmark-button"
          className="absolute top-1 right-1"
          onClick={() => setIsBookmarkedState((prevState) => !prevState)}
        >
          <StarIcon
            className={`${isListCard ? 'h-8 w-8' : 'h-6 w-6'}`}
            isFilled={isBookmarkedState}
          />
        </button>
      ) : null}
    </div>
  )
}

export default RepositoryCard
