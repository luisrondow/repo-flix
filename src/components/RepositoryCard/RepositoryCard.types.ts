export type Repository = {
  id: string
  name: string
  url: string
  image: string
}

export type RepositoryCardProps = {
  type: 'bookmark' | 'list'
  repository: Repository
  isBookmarked?: boolean
  onRepositoryClick: (url: string) => void
}
