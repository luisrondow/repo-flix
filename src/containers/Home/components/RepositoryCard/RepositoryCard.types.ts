export type RepositoryCardProps = {
  type: 'bookmark' | 'list'
  repository: {
    id: string
    name: string
    image: string
    url: string
  }
  onRepositoryClick: () => void
}
