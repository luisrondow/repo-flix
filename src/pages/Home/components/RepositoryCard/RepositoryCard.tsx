import { useState } from 'react'

import { RepositoryCardProps } from './RepositoryCard.types'
import StarIcon from '../../../../components/Icons/StarIcon'
import { useBookmarksContext } from '../../../../hooks/useBookmarksContext'

const RepositoryCard = (props: RepositoryCardProps) => {
  const { repository, onRepositoryClick, type } = props

  const { id, name, image } = repository

  const isListCard = type === 'list'

  const isMobile = window.innerWidth < 1280

  const { actions } = useBookmarksContext()

  const [isBookmarkedState, setIsBookmarkedState] = useState(
    () => actions?.isInBookmark(id) || !isListCard,
  )
  const [isCardHovered, setIsCardHovered] = useState(false)

  const handleBookmarkClick = () => {
    setIsBookmarkedState((prevState) => !prevState)

    if (isBookmarkedState) {
      actions?.removeFromBookmark(repository)
    }

    if (!isBookmarkedState) {
      actions?.addToBookmark(repository)
    }
  }

  return (
    <div
      data-testid={`repository-card-${id}`}
      className={`relative ${
        isListCard
          ? 'h-36 w-72 basis-72 sm:h-56 sm:w-112 sm:basis-112'
          : 'h-28 w-56 max-w-56 basis-56 sm:h-36 sm:w-72 sm:max-w-72 sm:basis-72'
      } flex-shrink-0 flex-grow cursor-pointer snap-center scroll-mx-96 transition-transform sm:hover:z-10 sm:hover:scale-115 sm:hover:shadow-md`}
      onMouseEnter={() => isListCard && setIsCardHovered(true)}
      onMouseLeave={() => isListCard && setIsCardHovered(false)}
    >
      <img src={image} alt={name} className="h-full w-full" onClick={onRepositoryClick} />
      {isCardHovered || !isListCard || isMobile ? (
        <button
          data-testid="bookmark-button"
          className="absolute top-1 right-1"
          onClick={handleBookmarkClick}
        >
          <StarIcon
            className={`${isListCard ? 'h-8 w-8' : 'h-6 w-6'}`}
            isFilled={(actions ? actions?.isInBookmark(id) : isBookmarkedState) || !isListCard}
          />
        </button>
      ) : null}
    </div>
  )
}

export default RepositoryCard
