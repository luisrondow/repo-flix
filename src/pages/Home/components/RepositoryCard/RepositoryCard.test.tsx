import { act, render, waitFor, fireEvent } from '@testing-library/react'

import RepositoryCard from './RepositoryCard'
import { RepositoryCardProps } from './RepositoryCard.types'

// jest.mock('../../../../hooks/useBookmarksContext', () => {
//   return {
//     __esModule: true,
//     default: () => ({
//       bookmarks: [],
//       actions: {
//         addBookmark: jest.fn(),
//         removeBookmark: jest.fn(),
//       },
//     }),
//   }
// })

describe('RepositoryCard', () => {
  const props: RepositoryCardProps = {
    repository: {
      id: 'id',
      name: 'mock',
      image: 'image://mock',
      url: 'url://mock',
    },
    type: 'list',
    onRepositoryClick: jest.fn(),
  }

  it('should render correctly', () => {
    const { container, getByRole } = render(<RepositoryCard {...props} />)

    expect(getByRole('img')).toHaveAttribute('src', props.repository.image)
    expect(container).toMatchSnapshot()
  })

  it('should render correctly when type is bookmark', () => {
    const { container, getByTestId } = render(<RepositoryCard {...props} type="bookmark" />)

    expect(getByTestId('bookmark-button')).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it('should call correctly the onRepositoryClick', () => {
    const { getByRole } = render(<RepositoryCard {...props} />)
    const card = getByRole('img')

    card.click()

    expect(props.onRepositoryClick).toHaveBeenCalled()
  })

  it('should render correctly the bookmark button', async () => {
    const { container, queryByTestId, getByTestId } = render(<RepositoryCard {...props} />)

    expect(queryByTestId('bookmark-button')).not.toBeInTheDocument()

    fireEvent.mouseEnter(getByTestId(`repository-card-${props.repository.id}`))

    await waitFor(() => getByTestId('bookmark-button'))

    expect(getByTestId('bookmark-button').closest('button')).toBeInTheDocument()

    expect(container).toMatchSnapshot()
  })

  it('should fire correctly the bookmark button click', async () => {
    const { container, getByTestId } = render(<RepositoryCard {...props} />)

    fireEvent.mouseEnter(getByTestId(`repository-card-${props.repository.id}`))

    await waitFor(() => getByTestId('bookmark-button'))

    const bookmark = getByTestId('bookmark-button')

    expect(bookmark.childNodes.item(0)).not.toHaveClass('fill-yellow')

    act(() => fireEvent.click(bookmark))

    await waitFor(() => {
      expect(bookmark.childNodes.item(0)).toHaveClass('fill-yellow')
    })

    expect(container).toMatchSnapshot()
  })
})
