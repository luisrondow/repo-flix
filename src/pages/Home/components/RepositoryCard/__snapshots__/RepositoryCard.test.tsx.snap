// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepositoryCard should fire correctly the bookmark button click 1`] = `
<div>
  <div
    class="relative h-36 w-72 basis-72 sm:h-56 sm:w-112 sm:basis-112 flex-shrink-0 flex-grow cursor-pointer snap-center scroll-mx-96 transition-transform sm:hover:z-10 sm:hover:scale-115 sm:hover:shadow-md"
    data-testid="repository-card-id"
  >
    <img
      alt="mock"
      class="h-full w-full"
      src="image://mock"
    />
    <button
      class="absolute top-1 right-1"
      data-testid="bookmark-button"
    >
      <svg
        class="h-8 w-8 fill-yellow stroke-yellow"
        fill="none"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M11.48 3.499a.562.562 0 011.04 0l2.125 5.111a.563.563 0 00.475.345l5.518.442c.499.04.701.663.321.988l-4.204 3.602a.563.563 0 00-.182.557l1.285 5.385a.562.562 0 01-.84.61l-4.725-2.885a.563.563 0 00-.586 0L6.982 20.54a.562.562 0 01-.84-.61l1.285-5.386a.562.562 0 00-.182-.557l-4.204-3.602a.563.563 0 01.321-.988l5.518-.442a.563.563 0 00.475-.345L11.48 3.5z"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`RepositoryCard should render correctly 1`] = `
<div>
  <div
    class="relative h-36 w-72 basis-72 sm:h-56 sm:w-112 sm:basis-112 flex-shrink-0 flex-grow cursor-pointer snap-center scroll-mx-96 transition-transform sm:hover:z-10 sm:hover:scale-115 sm:hover:shadow-md"
    data-testid="repository-card-id"
  >
    <img
      alt="mock"
      class="h-full w-full"
      src="image://mock"
    />
  </div>
</div>
`;

exports[`RepositoryCard should render correctly the bookmark button 1`] = `
<div>
  <div
    class="relative h-36 w-72 basis-72 sm:h-56 sm:w-112 sm:basis-112 flex-shrink-0 flex-grow cursor-pointer snap-center scroll-mx-96 transition-transform sm:hover:z-10 sm:hover:scale-115 sm:hover:shadow-md"
    data-testid="repository-card-id"
  >
    <img
      alt="mock"
      class="h-full w-full"
      src="image://mock"
    />
    <button
      class="absolute top-1 right-1"
      data-testid="bookmark-button"
    >
      <svg
        class="h-8 w-8 fill-none stroke-yellow"
        fill="none"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M11.48 3.499a.562.562 0 011.04 0l2.125 5.111a.563.563 0 00.475.345l5.518.442c.499.04.701.663.321.988l-4.204 3.602a.563.563 0 00-.182.557l1.285 5.385a.562.562 0 01-.84.61l-4.725-2.885a.563.563 0 00-.586 0L6.982 20.54a.562.562 0 01-.84-.61l1.285-5.386a.562.562 0 00-.182-.557l-4.204-3.602a.563.563 0 01.321-.988l5.518-.442a.563.563 0 00.475-.345L11.48 3.5z"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`RepositoryCard should render correctly when type is bookmark 1`] = `
<div>
  <div
    class="relative h-28 w-56 max-w-56 basis-56 sm:h-36 sm:w-72 sm:max-w-72 sm:basis-72 flex-shrink-0 flex-grow cursor-pointer snap-center scroll-mx-96 transition-transform sm:hover:z-10 sm:hover:scale-115 sm:hover:shadow-md"
    data-testid="repository-card-id"
  >
    <img
      alt="mock"
      class="h-full w-full"
      src="image://mock"
    />
    <button
      class="absolute top-1 right-1"
      data-testid="bookmark-button"
    >
      <svg
        class="h-6 w-6 fill-yellow stroke-yellow"
        fill="none"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M11.48 3.499a.562.562 0 011.04 0l2.125 5.111a.563.563 0 00.475.345l5.518.442c.499.04.701.663.321.988l-4.204 3.602a.563.563 0 00-.182.557l1.285 5.385a.562.562 0 01-.84.61l-4.725-2.885a.563.563 0 00-.586 0L6.982 20.54a.562.562 0 01-.84-.61l1.285-5.386a.562.562 0 00-.182-.557l-4.204-3.602a.563.563 0 01.321-.988l5.518-.442a.563.563 0 00.475-.345L11.48 3.5z"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
  </div>
</div>
`;
